// @social/social-deployment/templates/nodejs/api/BaseApi.js
/* eslint-disable func-names */
/* eslint-disable class-methods-use-this */

const zmq = require('zmq');
const ports = require('../../../topology/portMaps');

const serverTimeout = 2000;
const serverTimeoutMsg = {
    status: 408,
    message: 'Server timed out for request'
};
const serverErrorMsg = {
    status: 500,
    message: 'internal server error'
};
const malformedErrorMsg = {
    status: 500,
    message: 'malformed api call'
};

function makeMessage(ownerId, action, command, args = []) {
    let thisArgs = args;
    if (!Array.isArray(thisArgs)) thisArgs = [thisArgs];
    return JSON.stringify({
        ownerId,
        action,
        command,
        args: thisArgs
    });
}

function GetReqSocket(type) {
    this.type = type;
    this.socket = zmq.socket('req');
}

GetReqSocket.prototype.send = function (ownerId, action, command, args) {
    this.message = makeMessage(ownerId, action, command, args);
    return new Promise((resolve, reject) => {
        const socket = zmq.socket('req');
        socket.connect(`tcp://127.0.0.1:${ports[this.type].crud}`);
        socket.send(this.message);
        const timer = setTimeout(() => {
            socket.close();
            reject(serverTimeoutMsg);
        }, serverTimeout);
        socket.on('message', (msg) => {
            try {
                const m = JSON.parse(msg.toString());
                resolve(m);
            } catch (err) {
                console.log(err);
                reject(serverErrorMsg);
            } finally {
                clearTimeout(timer);
                socket.close();
            }
        });
    });
};


class BaseApi {
    constructor() {
        this.builder = (action, path, args = [], ownerId) => {
            let thisArgs = args;
            const route = path.split('.');
            if (route.length !== 2) return Promise.reject(malformedErrorMsg);
            const type = route[0];
            const command = route[1];
            if (!Array.isArray(thisArgs)) thisArgs = [thisArgs];
            const socket = this.getReqSocket(type);
            return socket.send(ownerId, action, command, args);
        };
        this.api = {
            create: (path, args, ownerId = null) => this.builder('create', path, args, ownerId),
            read: (path, args, ownerId = null) => this.builder('read', path, args, ownerId),
            update: (path, args, ownerId = null) => this.builder('update', path, args, ownerId),
            delete: (path, args, ownerId = null) => this.builder('delete', path, args, ownerId)
        };
    }

    getReqSocket(type) {
        return new GetReqSocket(type);
    }
}

module.exports = BaseApi;
